autonomi/src/client/high_level/register/mod.rs

```diff
@@ -0,0 +1,352 @@
+            .checked_add(pointer_cost?)
+            .ok_or(CostError::InvalidCost)
+    }
+
+    /// Get the address of the register's head pointer
+    fn register_head_pointer_address(&self, addr: &RegisterAddress) -> PointerAddress {
+        let pk: MainPubkey = addr.owner.into();
+        let pointer_pk =
+            pk.derive_key(&DerivationIndex::from_bytes(REGISTER_HEAD_DERIVATION_INDEX));
+        PointerAddress::from_owner(pointer_pk.into())
+    }
+
+    /// Get the secret key of the register's head pointer
+    fn register_head_pointer_sk(&self, register_owner: &SecretKey) -> SecretKey {
+        let pointer_sk = MainSecretKey::new(register_owner.clone())
+            .derive_key(&DerivationIndex::from_bytes(REGISTER_HEAD_DERIVATION_INDEX));
+        pointer_sk.into()
+    }
+
+    /// Get the public key of the register's head pointer
+    fn register_head_pointer_pk(&self, addr: &RegisterAddress) -> PublicKey {
+        let pk: MainPubkey = addr.owner.into();
+        let pointer_pk =
+            pk.derive_key(&DerivationIndex::from_bytes(REGISTER_HEAD_DERIVATION_INDEX));
+        pointer_pk.into()
+    }
+
+    /// Get underlying register graph entry and next derivation index
+    /// In normal circumstances, there is only one entry with one descendant, yielding ONE entry and ONE derivation index
+    /// In the case of a fork or a corrupt register, the smallest derivation index among all the entries descendants is chosen
+    /// We chose here to deal with the errors instead of erroring out to allow users to solve Fork and Corrupt issues by updating the register
+    async fn register_get_graph_entry_and_next_derivation_index(
+        &self,
+        graph_entry_addr: &GraphEntryAddress,
+    ) -> Result<(GraphEntry, DerivationIndex), RegisterError> {
+        let entry = match self.graph_entry_get(*graph_entry_addr).await {
+            Ok(e) => e,
+            Err(GraphError::Fork(entries)) => {
+                warn!("Forked register, multiple entries found: {entries:?}, choosing the one with the smallest derivation index for the next entry");
+                let (entry_by_smallest_derivation, _) = entries
+                    .into_iter()
+                    .filter_map(|e| {
+                        e.descendants
+                            .iter()
+                            .map(|d| d.1)
+                            .min()
+                            .map(|derivation| (e, derivation))
+                    })
+                    .min_by(|a, b| a.1.cmp(&b.1))
+                    .ok_or(RegisterError::Corrupt(format!(
+                        "No descendants found for FORKED entry at {graph_entry_addr:?}"
+                    )))?;
+                entry_by_smallest_derivation
+            }
+            Err(err) => return Err(err.into()),
+        };
+        let new_derivation =
+            entry
+                .descendants
+                .iter()
+                .map(|d| d.1)
+                .min()
```

Comment:
For Register, if have multiple descendants, shall raise error instead ?

Reply:
Yeah maybe that would be good indeed. Agreed I'll make an error here